### ISSUE TITLE
dotnet-sdk DEB and RPM package consume store by "dependency" field

### ISSUE THREAD

**Issue by wli3** [dotnet/cli#7154]
The method of getting runtime package store should be changed from being part of dotnet-sdk package to obtain via native package manager which will download and install the runtime package store.

The dependency graph should be (Updated awesome ascii art)
```
                              +------------------+   +------------------+    +--------------+
                 +------------>   dotnet-runtime +--->  dotnet-hostfxr  +---->dotnet-host   |
                 |            |                  |   |                  |    |              |
                 |            +------------------+   +------------------+    +--------------+
+------------+   |
| dotnet-sdk +---+
|            +---+            +------------------+
+------------+   |            |aspnetcore-store  |
                 +------------>                  |
                              +------------------+


```

**Comment by omajid** [comment]
Shouldn't it be the runtime package that depends on the store? The SDK doesn't use the store, does it? If the user only installs the runtime, they still need the store too, no?

**Comment by wli3** [comment]
@JunTaoLuo @Petermarcu Please correct me if I am wrong

> runtime package that depends on the store

no, runtime don't depend on store

> The SDK doesn't use the store,

Not directly, more detail is here https://github.com/dotnet/designs/issues/8

> If the user only installs the runtime, they still need the store too, no?

as the first one, it is not required

**Comment by Petermarcu** [comment]
I would expect people go an install the aspnetcore-store package which would depend on dotnet-runtime.

**Comment by Petermarcu** [comment]
Another thing we had talked about having an aspnet-hosting package that pulled in the dotnet-runtime and aspnetcore-store on their own and not having the aspnetcore-store actually depend on anything.

**Comment by Petermarcu** [comment]
@omajid the store is not required for .NET Core apps. It is only required for ASP.NET Core apps.

**Comment by omajid** [comment]
I think the actual dependency chain - that works in practice -  is completely upside down from what is  intuitive to everyone. I was confused at first too. Here's what I understand about the *runtime* package store now.

See https://github.com/dotnet/designs/issues/8#issuecomment-308593958 for a developer clarifying that:

> it's a runtime only feature

In other words, the SDK does *not* depend on it. The SDK does not use it. You can test it yourself. Get an SDK, delete the store and then use it to build ASP.NET Core applications.

By default, if the store it is missing, your ASP.NET Core  applications will mysteriously fail to run (only when published as a framework-dependent application, as I understand it). I say "by default" because there are workarounds to change that.

See also https://github.com/dotnet/designs/issues/10 which points out the affect of missing runtime store is in the runtime environment, not the developer environment:

> When an ASP.NET published application gets deployed, one should make sure that the target environment has ASP.NET installed, as the presence of .NET Core alone will not be sufficient.

In other words:

- SDK does not need the store
- Runtime works without the store, unless the application is ASP.NET Core (that is, it actually tries to use the store - console applications don't use it). 
- If the store is needed, the runtime fails to run the application with fairly obscure messages.

Seems to me like it would be more appropriate to say the "runtime" package depends on the "store" since without the store the runtime fails.

**Comment by Petermarcu** [comment]
The ASP.NET "runtime" package needs to depend on the store. That would be the the aspnet-hosting package I mentioned. ASP.NET is going to make that package available and docs will tell people building ASP.NET apps to install that to satisfy runtime dependencies. 

Apps that are not asp.net apps, do not need the store installed.

**Comment by omajid** [comment]
Oh. I see.

In that case, it sounds like the dependency doesn't belong here?

**Comment by Petermarcu** [comment]
The SDK does depend on it because we want ASP.NET apps to be able to run by default when the SDK is installed. Then at runtime, people need to install the aspnet-hosting package to get all the aspnet dependencies.

**Comment by julielerman** [comment]
I'm on debian 9 and trying to install .NET Core 2.0.X daily build. I initially had an issue because runtime dependended on outdated versions of some files (libssl 1.0.0, i had 1.0.2 and 1.1.0 as well as libicu52 and I have libicu57). I installed those but I'm still getting an issue with aspnetcore-store.2.0.0:

**Comment by omajid** [comment]
@julielerman You have added the apt feed, right? Can you try using `apt-get install dotnet-sdk-latest`?

**Comment by julielerman** [comment]
I could try another time but I literally downloaded and installed the missing debian packages and was able to move forward. And I can't go back and experiment right now. But I downloaded the latest deb packages right from this repo. I wasn't using apt-get to install dotnet or the sdk.

**Comment by wli3** [comment]
@julielerman there will be a debian.9 runtime for debian 9. @rakeshsinghranchi for more detail

**Comment by omajid** [comment]
> I could try another time but I literally downloaded and installed the missing debian packages and was able to move forward.

Unfortunately, that is pretty much how `dpkg` works by design. You need to have all the `.deb` files on disk - including all recursive dependencies - and then `dpkg` will install them. `apt` sits on top of `dpkg` and will download `.deb`s from the apt feeds (including recursive dependencies) and will then install them.

**Comment by julielerman** [comment]
Thanks @wli3 and @omajid . This was my first ever experience with Linux so I didn't even realize that `dpkg `works like `nuget `...going back to the feed to get other dependencies.  (yeah ...DUH!!!) When I get out of my hole, I will try it again because it will make me feel better to have done it properly.

**Comment by omajid** [comment]
Sorry, just a quick clarification. `apt` (and `apt-get`, but also `yum` and `dnf` in other Linuxes) works like nuget: downloading and installing recursive dependencies. `dpkg` (and `rpm`) don't. `dpkg` (and `rpm`) are considered lower level tools used when there is a good reason to. Most users would default to using `apt` (or `apt-get`, or `yum`, or `dnf`, or ...). You should try `apt`!

**Comment by julielerman** [comment]
ahh okay. I do use apt-get for a ton of things, but I somehow thought I was safer with the debs for this. Well, I have a feeling there will be more linux installs in my future to practice this on. :) Since I was running apt-get update -f after dpkg, should I have been able to ignore the package warnings? (I'm sorry , I know I can test this theory out ...)
